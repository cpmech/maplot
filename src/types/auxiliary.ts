export type ICoordsToString = (x: number, y: number, xScreen?: number, yScreen?: number) => string;
