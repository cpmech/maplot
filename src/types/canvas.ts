export interface IPalettes {
  [name: string]: string[];
}

export type LineStyle = '-' | '--' | ':' | 'none';
