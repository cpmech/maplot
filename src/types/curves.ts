export type LineStyle = '-' | '--' | ':' | 'none';

export type MarkerType = 'o' | 's' | '+' | 'x' | 'img' | 'none';

export interface ICurveStyle {
  // lines
  lineColor: string; // color
  lineAlpha: number; // alpha (0, 1]. A<1e-14 => A=1.0
  lineStyle: LineStyle; // style
  lineWidth: number; // width

  // markers
  markerType: MarkerType; // type
  markerImg: string; // image filename
  markerColor: string; // color
  markerAlpha: number; // alpha (0, 1]
  markerSize: number; // size; when using images, set markerSize=0 to use the image width
  markerEvery: number; // mark-every
  markerLineColor: string; // edge color
  markerLineWidth: number; // edge width
  markerLineStyle: LineStyle; // edge style
  markerIsVoid: boolean; // void marker (draw edge only)
}

export interface ICurve {
  style: ICurveStyle; // line and marker arguments
  label: string; // curve name or connection pair such as 'San Francisco -> Los Angeles'
  x: number[]; // x-coordinates
  y: number[]; // y-coordinates
  z?: number[]; // z-coordinates
  kind?: string; // e.g. connection, city, fortress, base, mine, ...
  tagFirstPoint?: boolean; // tag first point with label
}

export interface ICurves {
  list: ICurve[];
}

export const defaultCurveStyle: ICurveStyle = {
  // lines
  lineColor: '#b33434',
  lineAlpha: 0.7,
  lineStyle: '-',
  lineWidth: 3,

  // markers
  markerType: 'o',
  markerImg: '',
  markerColor: '#4c4deb',
  markerAlpha: 1,
  markerSize: 0,
  markerEvery: 0,
  markerLineColor: '#ffffff',
  markerLineWidth: 2,
  markerLineStyle: 'none',
  markerIsVoid: false,
};

export const newZeroCurveStyle = (): ICurveStyle => ({
  lineColor: '',
  lineAlpha: 0,
  lineStyle: '-',
  lineWidth: 0,
  markerType: '+',
  markerImg: '',
  markerColor: '',
  markerAlpha: 0,
  markerSize: 0,
  markerEvery: 0,
  markerLineColor: '',
  markerLineWidth: 0,
  markerLineStyle: '-',
  markerIsVoid: false,
});

export const newZeroCurve = (): ICurve => ({
  style: newZeroCurveStyle(),
  label: '',
  x: [],
  y: [],
  z: [],
  kind: '',
  tagFirstPoint: false,
});
