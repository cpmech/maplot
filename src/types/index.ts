export * from './canvas';
export * from './curves';
export * from './geometry';
export * from './plotargs';
export * from './resizer';
export * from './ticksinfo';
