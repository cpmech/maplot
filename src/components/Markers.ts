import { drawCircle, setStroke, drawRect, drawLine } from '../canvas';
import { loadImages } from '../io';
import { hasProp } from '../helpers';
import { ICurveStyle, IPlotArgs } from '../types';

export class Markers {
  private dc: CanvasRenderingContext2D;
  private args: IPlotArgs;
  private images: { [filePath: string]: HTMLImageElement } = {};

  constructor(dc: CanvasRenderingContext2D, args: IPlotArgs) {
    this.dc = dc;
    this.args = args;
  }

  async init(): Promise<void> {
    if (this.args.markerImgPaths.length > 0) {
      const imgs = await loadImages(this.args.markerImgPaths);
      imgs.forEach((img, index) => {
        this.images[this.args.markerImgPaths[index]] = img;
      });
    }
  }

  getSize(style: ICurveStyle, referenceWidth: number): number {
    if (style.markerType === 'img' && style.markerSize === 0) {
      const img = this.img(style);
      return this.size(img.width, referenceWidth);
    }
    return this.size(style.markerSize, referenceWidth);
  }

  draw(x: number, y: number, style: ICurveStyle, referenceWidth: number) {
    // 'none'
    if (style.markerType === 'none') {
      return;
    }

    // 'img'
    if (style.markerType === 'img') {
      const img = this.img(style);
      const msize = style.markerSize === 0 ? img.width : style.markerSize;
      const s = this.size(msize, referenceWidth);
      const h = s / 2;
      this.dc.drawImage(img, x - h, y - h, s, s);
      return;
    }

    // 'o'
    if (style.markerType === 'o') {
      const r = this.size(style.markerSize, referenceWidth);
      this.activateStyle(style);
      if (style.markerIsVoid) {
        drawCircle(this.dc, x, y, r, false);
      } else {
        drawCircle(this.dc, x, y, r, true);
        if (style.markerLineColor && style.markerLineColor !== style.markerColor) {
          setStroke(
            this.dc,
            style.markerLineColor,
            style.markerAlpha,
            style.markerLineWidth,
            style.markerLineStyle,
          );
          drawCircle(this.dc, x, y, r, false);
        }
      }
      return;
    }

    // 's'
    if (style.markerType === 's') {
      const w = this.size(style.markerSize, referenceWidth);
      this.activateStyle(style);
      if (style.markerIsVoid) {
        drawRect(this.dc, x, y, w, w, false);
      } else {
        drawRect(this.dc, x, y, w, w, true);
        if (style.markerLineColor && style.markerLineColor !== style.markerColor) {
          setStroke(
            this.dc,
            style.markerLineColor,
            style.markerAlpha,
            style.markerLineWidth,
            style.markerLineStyle,
          );
          drawRect(this.dc, x, y, w, w, false);
        }
      }
      return;
    }

    // '+'
    if (style.markerType === '+') {
      const h = this.size(style.markerSize, referenceWidth);
      this.activateStyle(style);
      drawLine(this.dc, x - h, y, x + h, y);
      drawLine(this.dc, x, y - h, x, y + h);
      return;
    }

    // 'x'
    if (style.markerType === 'x') {
      const h = this.size(style.markerSize, referenceWidth);
      this.activateStyle(style);
      drawLine(this.dc, x - h, y - h, x + h, y + h);
      drawLine(this.dc, x - h, y + h, x + h, y - h);
      return;
    }

    throw new Error(`markerType = "${style.markerType}" is invalid`);
  }

  private img(style: ICurveStyle) {
    if (!hasProp(this.images, style.markerImg)) {
      throw new Error(`cannot find image named ${style.markerImg}. did you forget to call init()?`);
    }
    return this.images[style.markerImg];
  }

  private size(markerSize: number, referenceWidth: number): number {
    const mm = this.args.markerSizeAuto ? this.args.markerSizeRefProp * referenceWidth : 1.0;
    const ms = markerSize === 0 ? this.args.markerSizeDefault : markerSize;
    const sz = mm * ms;
    if (sz < this.args.markerSizeMin) {
      return this.args.markerSizeMin;
    }
    if (sz > this.args.markerSizeMax) {
      return this.args.markerSizeMax;
    }
    return sz;
  }

  private activateStyle(style: ICurveStyle) {
    if (style.markerIsVoid) {
      setStroke(
        this.dc,
        style.markerColor,
        style.markerAlpha,
        style.markerLineWidth,
        style.markerLineStyle,
      );
      return;
    }
    this.dc.fillStyle = style.markerColor;
    this.dc.strokeStyle = style.markerColor;
    this.dc.globalAlpha = style.markerAlpha;
    this.dc.lineWidth = style.markerLineWidth;
  }
}
