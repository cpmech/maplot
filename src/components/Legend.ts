import { IPlotArgs, ICurve, ICurves, ICurveStyle } from '../types';
import { setStroke, drawLine, drawRect, drawFilledRectWithEdge, textWidthPx } from '../canvas';
import { Markers } from './Markers';

export class Legend {
  // input
  dc: CanvasRenderingContext2D;
  args: IPlotArgs;
  curves: ICurves;
  markers: Markers;

  constructor(dc: CanvasRenderingContext2D, args: IPlotArgs, curves: ICurves, markers: Markers) {
    this.dc = dc;
    this.args = args;
    this.curves = curves;
    this.markers = markers;
  }

  getDims(): { width: number; height: number } {
    return {
      width: 600,
      height: 600,
    };
  }

  render(offsetX: number, offsetY: number) {
    this.dc.save();
    this.dc.translate(offsetX, offsetY);

    if (this.curves.list.length < 1) {
      return;
    }

    const { width, height } = this.getDims();

    drawFilledRectWithEdge(this.dc, 0, 0, width, height);

    const curves = this.curves.list;
    const l = this.args.l;
    const nCol = l.nCol > 0 ? l.nCol : l.atBottom ? curves.length : 1;
    const nRow = Math.trunc(curves.length / nCol) + 1;

    console.log('nRow =', nRow, ' nCol =', nCol);

    const colWidths = Array(nCol).fill(0);
    const rowHeights = Array(nRow).fill(0);
    for (let iCurve = 0; iCurve < curves.length; iCurve++) {
      const row = Math.floor(iCurve / nCol);
      const col = iCurve % nCol;
      const { w, h } = this.iconDims(curves[iCurve]);
      colWidths[col] = Math.max(colWidths[col], w);
      rowHeights[row] = Math.max(rowHeights[row], h);
    }
    console.log('colWidths =', colWidths, ' rowHeights =', rowHeights);

    let y = 0;
    let idxCurve = 0;
    for (let row = 0; row < nRow; row++) {
      let x = 0;
      for (let col = 0; col < nCol; col++) {
        const curve = curves[idxCurve];
        this.iconTextAtRight(x, y, colWidths[col], rowHeights[row], curve.style);
        x += colWidths[col] + l.gapIconsHoriz;
        idxCurve++;
        if (idxCurve === curves.length) {
          break;
        }
      }
      if (idxCurve === curves.length) {
        break;
      }
      y += rowHeights[row] + l.gapIconsVert;
    }

    this.dc.restore();
  }

  private iconDims(curve: ICurve): { w: number; h: number } {
    // const lineLen = Math.max(this.args.l.lineLen, style.markerSize)
    const font = `${this.args.fsizeLegend}px ${this.args.fnameLegend}`;
    const l = this.args.l;
    const labelLen = textWidthPx(this.dc, curve.label, font);
    return {
      w: l.lineLen + l.gapLineLabel + labelLen,
      h: Math.max(curve.style.markerSize, this.args.fsizeLegend),
    };
  }

  private iconTextAtRight(x: number, y: number, w: number, h: number, style: ICurveStyle) {
    drawFilledRectWithEdge(this.dc, x, y, w, h, '#cecece', '#ececec');

    // const lineLen = 20;
    // const gapLeft = 25;
    // const gapMiddle = 5;
    // const gapRight = 50;
    //
    //            |← lineLen →|← gapMiddle →|← labelLen →|
    //   gapLeft →[   line    |             |    txt     ]← gapRight
    //
    //  example:         ——o—— x = y²
    //

    // const u = u0 + gapLeft;
    // const v = v0;
    // const du = lineLen;

    const l = this.args.l;
    if (style.lineStyle !== 'none') {
      const dx = l.lineLen;
      const dy = h / 2;
      setStroke(this.dc, style.lineColor, style.lineAlpha, style.lineWidth, style.lineStyle);
      drawLine(this.dc, x, y + dy, x + dx, y + dy);
    }
  }
}
