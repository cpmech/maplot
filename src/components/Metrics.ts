import { IPlotArgs, ICurves, ILimits } from '../types';
import { cteEps } from './constants';
import { Markers } from './Markers';
import { Legend } from './Legend';
import { Scale } from './Scale';
import { Ticks } from './Ticks';
import { Limits, getLimitsAroundCurves, checkLimitsForConsistency } from './Limits';

//                |←———————————————————————— W ————————————————————————————————→|
//                |←——— LR ———|←——————————— ww ——————————→|←——————— RR ————————→|
//                            .    |←—————— w ——————→|    .                     .
//              (0,0)         .                           .                     .
//   ———————————  ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓ ——→ xScr
//    ↑   ↑       ┃                         gTR                                 ┃
//    |  TR       ┃                                                             ┃
//    |   ↓       ┃        (x0,y0)          gTR                                 ┃
//    | ————————  ┃  . . . .  ┌───────────────────────────┐        —————        ┃ ——→ x
//    |   ↑       ┃           │ (p0,q0)     ' dV          │          ↑          ┃
//    |   |  ———  ┃           │    o─────────────────┐    │          |          ┃ ——→ p
//    |   |   ↑   ┃           │    │                 │    │    |     |    |     ┃
//    |           ┃  gLR | LS │-dH-│                 │-dH-│-RS-|-gRR-|-RL-|-gRR-┃
//       hh   h   ┃           │    │                 │    │    |     |    |     ┃
//    H           ┃           │    ↑ yReal           │    │          |          ┃
//        |   ↓   ┃           │    │                 │    │          |          ┃
//    |   |  ———  ┃           │    ●─────────────────o    │          |          ┃ ——→ xReal
//    |   ↓       ┃           │             ' dV  (pf,qf) │          ↓          ┃
//    |  ———————  ┃  . . . .  └───────────────────────────┘        —————        ┃
//    |   ↑       ┃                        BS          (xf,yf)                  ┃
//    |   |       ┃                      ───────                                ┃
//    |           ┃                        gBR                                  ┃
//    |  BR       ┃                      ───────                                ┃
//    |           ┃                        BL                                   ┃
//    |   |       ┃                      ───────                                ┃
//    ↓   ↓       ┃                        gBR                                  ┃
//   ———————————  ┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
//                |           |    |                                          (W,H)
//                ↓           ↓    ↓
//               yScr         y    q
//
//    W: figure width          ww: plot area width     w: inner x-y area width
//    H: figure height         hh: plot area height    h: inner x-y area height
//
//    LR : Left Ruler          TR : Top Ruler          xScr ,yScr : "screen" coordinates
//    LS : Left Scale          gLR: gap for LR         xReal,yReal: x-y (real data) coords
//    BR : Bottom Ruler        RR : Right Ruler        x    ,y    : plot area coordinates
//    BL : Bottom Legend       RL : Right Legend
//    BS : Bottom Scale        RS : Right Scale
//    gRR: gap for RS          gBR: gap for BS
//    dH : Delta Horizontal    dV : Delta Vertical

export class Metrics {
  // input
  dc: CanvasRenderingContext2D;
  args: IPlotArgs;
  curves: ICurves;
  markers: Markers;
  legend?: Legend;

  // scale, limits and ticks
  xscale: Scale;
  yscale: Scale;
  limits: Limits;
  ticks: Ticks;

  // marker multiplier
  markerRefLength: number = 800; // reference width to update marker sizes upon zoom
  curveLimits: ILimits = { xmin: -1, xmax: +1, ymin: -1, ymax: +1 }; // reference values to set markers

  // geometry
  W: number = 0; // figure width
  H: number = 0; // figure height
  LR: number = 0; // Left Ruler thickness (screen coordinates)
  RR: number = 0; // Right Ruler thickness (screen coordinates)
  BR: number = 0; // Bottom Ruler thickness (screen coordinates)
  TR: number = 0; // Top Ruler thickness (screen coordinates)
  BL: number = 0; // Bottom Legend thickness
  RL: number = 0; // Right Legend thickness
  BS: number = 0; // Bottom Scale
  LS: number = 0; // Left Scale
  RS: number = 0; // Right Scale
  ww: number = 0; // width of plotting area
  hh: number = 0; // height of plotting area
  w: number = 0; // width of inner x-y area
  h: number = 0; // height of inner x-y area

  // scaling factors
  sfx: number = 0; // x scale factor
  sfy: number = 0; // y scale factor

  // coordinates
  x0: number = 0; // x-origin of plotting area
  y0: number = 0; // y-origin of plotting area
  xf: number = 0; // x-max of plotting area
  yf: number = 0; // y-max of plotting area
  p0: number = 0; // x-origin of inner x-y area
  q0: number = 0; // y-origin of inner x-y area
  pf: number = 0; // x-max of plotting area
  qf: number = 0; // y-max of plotting area

  constructor(
    dc: CanvasRenderingContext2D,
    args: IPlotArgs,
    curves: ICurves,
    markers: Markers,
    legend?: Legend,
  ) {
    this.dc = dc;
    this.args = args;
    this.curves = curves;
    this.markers = markers;
    this.legend = legend;
    this.xscale = new Scale(dc, args, this, false);
    this.yscale = new Scale(dc, args, this, true);
    this.limits = new Limits(args, curves);
    this.ticks = new Ticks(args, this.limits);
  }

  resize(width: number, height: number) {
    if (this.curves.list.length > 0) {
      this.setScaleBasedOnCurves(width, height);
    } else {
      this.setScaleFromStartValues(width, height);
    }
  }

  setScaleBasedOnCurves(width: number, height: number) {
    if (this.curves.list.length > 0) {
      this.W = width;
      this.H = height;
      this.curveLimits = this.limits.aroundCurves();
      this.ticks.calculate();
      this.calculateMetrics();
      this.expandLimitsDueToEqualScale();
      this.smallShiftToCurves();
      this.limits.makeCopy();
      this.calcMarkerMultiplier();
    }
  }

  setScaleFromStartValues(width: number, height: number) {
    if (
      checkLimitsForConsistency(
        this.args.x.minStart,
        this.args.x.maxStart,
        this.args.y.minStart,
        this.args.y.maxStart,
      )
    ) {
      this.W = width;
      this.H = height;
      this.limits.xmin = this.args.x.minStart;
      this.limits.xmax = this.args.x.maxStart;
      this.limits.ymin = this.args.y.minStart;
      this.limits.ymax = this.args.y.maxStart;
      this.ticks.calculate();
      this.calculateMetrics();
      this.expandLimitsDueToEqualScale();
      this.limits.makeCopy();
      this.calcMarkerMultiplier();
    }
  }

  copyLimits() {
    this.limits.makeCopy();
  }

  translateLimitsRelativeToCopy(dx: number, dy: number) {
    this.limits.translateRelativeToCopy(dx, dy);
    this.ticks.calculate();
    this.calculateMetrics();
  }

  zoom(factor: number) {
    const dxOld = this.limits.xmax - this.limits.xmin;
    const dyOld = this.limits.ymax - this.limits.ymin;
    const dxNew = dxOld * factor;
    const dyNew = dyOld * factor;
    const dx = (dxOld - dxNew) / 2.0;
    const dy = (dyOld - dyNew) / 2.0;
    if (
      checkLimitsForConsistency(
        this.limits.xmin + dx,
        this.limits.xmax - dx,
        this.limits.ymin + dy,
        this.limits.ymax - dy,
      )
    ) {
      this.limits.xmin += dx;
      this.limits.xmax -= dx;
      this.limits.ymin += dy;
      this.limits.ymax -= dy;
      this.ticks.calculate();
      this.calculateMetrics();
      this.limits.makeCopy();
      this.calcMarkerMultiplier();
      return true;
    }
  }

  translateTo(x: number, y: number) {
    this.limits.translateTo(x, y);
    this.ticks.calculate();
    this.calculateMetrics();
    this.limits.makeCopy();
  }

  // xScr converts real x-coords to to screen coordinates
  xScr(x: number): number {
    if (this.args.x.invert) {
      return Math.trunc(this.pf - this.sfx * (x - this.limits.xmin));
    } else {
      return Math.trunc(this.p0 + this.sfx * (x - this.limits.xmin));
    }
  }

  // yScr converts real y-coords to to screen coordinates
  yScr(y: number): number {
    if (this.args.y.invert) {
      return Math.trunc(this.q0 + this.sfy * (y - this.limits.ymin));
    } else {
      return Math.trunc(this.qf - this.sfy * (y - this.limits.ymin));
    }
  }

  // xReal converts to real coordinates
  xReal(xscr: number) {
    if (this.args.x.invert) {
      return this.limits.xmin + (this.pf - xscr) / this.sfx;
    } else {
      return this.limits.xmin + (xscr - this.p0) / this.sfx;
    }
  }

  // xReal converts to real coordinates
  yReal(yscr: number) {
    if (this.args.y.invert) {
      return this.limits.ymin + (yscr - this.q0) / this.sfy;
    } else {
      return this.limits.ymin + (this.qf - yscr) / this.sfy;
    }
  }

  // dxReal converts screen increment to real coordinates
  dxReal(dxscr: number) {
    return dxscr / this.sfx;
  }

  // dyReal converts screen increment to real coordinates
  dyReal(dyscr: number) {
    return dyscr / this.sfy;
  }

  private calculateMetrics() {
    // height of title
    let titleHeightPlusGap = 0;
    if (this.args.title) {
      titleHeightPlusGap = this.args.fsizeTitle + 2 * this.args.gTR;
    }

    // scales
    this.BS = this.xscale.getSize();
    this.LS = this.yscale.getSize();
    this.RS = 0;

    // legend dimensions
    this.BL = 0;
    this.RL = 0;
    let gBRtotal = this.args.gBR;
    let gRRtotal = this.args.gRR;
    if (this.legend && this.args.l.on) {
      const legDims = this.legend.getDims();
      if (this.args.l.atBottom) {
        gBRtotal += this.args.gBR;
        this.BL = legDims.height;
      } else {
        gRRtotal = this.args.gRR;
        this.RL = legDims.width;
      }
    }

    // rulers
    this.TR = Math.max(this.args.minTR, titleHeightPlusGap);
    this.LR = Math.max(this.args.minLR, this.args.gLR + this.LS);
    this.BR = Math.max(this.args.minBR, gBRtotal + this.BS + this.BL);
    this.RR = Math.max(this.args.minRR, gRRtotal + this.RS + this.RL);

    // geometry
    this.ww = Math.max(1, this.W - this.LR - this.RR);
    this.hh = Math.max(1, this.H - this.TR - this.BR);
    this.w = Math.max(1, this.ww - 2 * this.args.dH);
    this.h = Math.max(1, this.hh - 2 * this.args.dV);

    // coordinates
    this.x0 = this.LR;
    this.y0 = this.TR;
    this.p0 = this.x0 + this.args.dH;
    this.q0 = this.y0 + this.args.dV;
    this.xf = this.x0 + this.ww;
    this.yf = this.y0 + this.hh;
    this.pf = this.p0 + this.w;
    this.qf = this.q0 + this.h;

    // scale factors
    this.sfx = this.w / (this.limits.xmax - this.limits.xmin);
    this.sfy = this.h / (this.limits.ymax - this.limits.ymin);
    if (this.sfx <= cteEps) {
      this.sfx = 1.0;
    }
    if (this.sfy <= cteEps) {
      this.sfy = 1.0;
    }
    if (this.args.equalScale) {
      const sf = Math.min(this.sfx, this.sfy);
      this.sfx = sf;
      this.sfy = sf;
    }
  }

  private expandLimitsDueToEqualScale() {
    if (!this.args.equalScale) {
      return;
    }

    //  xmin       xmaxTmp    xmaxNew
    //    ____________  _ _ _ _ _
    //   |            |          |  ymax
    //   |                       |
    //   | dy         |          .
    //   |                       |
    //   |  dx := dy  |          .
    //   |____________  _ _ _ _ _|  ymin

    // expand limits along larger dimension
    if (this.w > this.h) {
      if (this.args.x.invert) {
        this.limits.xmax = this.xReal(this.x0);
      } else {
        this.limits.xmax = this.xReal(this.xf);
      }
    } else {
      if (this.args.y.invert) {
        this.limits.ymax = this.yReal(this.yf);
      } else {
        this.limits.ymax = this.yReal(this.y0);
      }
    }
  }

  private smallShiftToCurves() {
    if (this.curves.list.length > 0) {
      const res = getLimitsAroundCurves(this.args, this.curves);
      const xmid = (res.xmin + res.xmax) / 2;
      const ymid = (res.ymin + res.ymax) / 2;
      const ucen = (this.p0 + this.pf) / 2;
      const vcen = (this.q0 + this.qf) / 2;
      const xcen = this.xReal(ucen);
      const ycen = this.yReal(vcen);
      const dx = xcen - xmid;
      const dy = ycen - ymid;
      this.limits.xmin -= dx;
      this.limits.xmax -= dx;
      this.limits.ymin -= dy;
      this.limits.ymax -= dy;
      this.ticks.decreaseLo(dx, dy);
    }
  }

  private calcMarkerMultiplier() {
    this.markerRefLength = Math.min(this.w, this.h);
  }
}
