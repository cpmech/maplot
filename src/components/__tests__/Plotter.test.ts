import { IPlotArgs, ICurves, defaultPlotArgs, defaultCurveStyle } from '../../types';
import { Markers } from '../Markers';
import { Legend } from '../Legend';
import { Metrics } from '../Metrics';
import { Plotter } from '../Plotter';

let canvas: HTMLCanvasElement;
let dc: CanvasRenderingContext2D;

beforeAll(() => {
  canvas = document.createElement('canvas');
  const theDC = canvas.getContext('2d');
  if (theDC) {
    dc = theDC;
    return;
  }
  fail('internal error');
});

const args: IPlotArgs = { ...defaultPlotArgs };

const curves: ICurves = {
  list: [
    {
      label: 'First Curve',
      kind: 'Group A',
      x: [0.0, 0.0, 0.5, 1.0, 1.0, 0.0, 1.0, 0.3, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
      y: [0.0, 1.0, 1.0, 0.7, 0.3, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 0.8, 0.6, 0.4, 0.4],
      z: [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.2, 0.5, 0.8, 1.0, 1.0, 0.5, 0.0, 0.0, 0.5, 1.0],
      style: { ...defaultCurveStyle },
      tagFirstPoint: false,
    },
  ],
};

describe('Plotter', () => {
  const markers = new Markers(dc, args);
  const legend = new Legend(dc, args, curves, markers);

  it('constructor works', () => {
    const metrics = new Metrics(dc, args, curves, markers, legend);
    const plt = new Plotter(dc, args, curves, markers, legend, metrics);
    expect(plt.curves.list.length).toBe(1);
  });

  it('default render works', () => {
    const metrics = new Metrics(dc, args, curves, markers, legend);
    const plt = new Plotter(dc, args, curves, markers, legend, metrics);
    plt.render();
    expect(dc).toMatchSnapshot();
  });
});
