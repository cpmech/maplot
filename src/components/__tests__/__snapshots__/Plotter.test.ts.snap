// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plotter default render works 1`] = `
CanvasRenderingContext2D {
  "_canvas": <canvas />,
  "_directionStack": Array [
    "inherit",
    "inherit",
  ],
  "_drawCalls": Array [
    Object {
      "props": Object {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
      "transform": Array [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "fillRect",
    },
  ],
  "_events": Array [
    Object {
      "props": Object {},
      "transform": Array [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "save",
    },
    Object {
      "props": Object {
        "x": 0,
        "y": 0,
      },
      "transform": Array [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "translate",
    },
    Object {
      "props": Object {
        "value": 1,
      },
      "transform": Array [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "globalAlpha",
    },
    Object {
      "props": Object {
        "value": "#fff",
      },
      "transform": Array [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "fillStyle",
    },
    Object {
      "props": Object {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
      "transform": Array [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "fillRect",
    },
  ],
  "_fillStyleStack": Array [
    "#000",
    "#fff",
  ],
  "_filterStack": Array [
    "none",
    "none",
  ],
  "_fontStack": Array [
    "10px sans-serif",
    "10px sans-serif",
  ],
  "_globalAlphaStack": Array [
    1,
    1,
  ],
  "_globalCompositeOperationStack": Array [
    "source-over",
    "source-over",
  ],
  "_imageSmoothingEnabledStack": Array [
    true,
    true,
  ],
  "_imageSmoothingQualityStack": Array [
    "low",
    "low",
  ],
  "_lineCapStack": Array [
    "butt",
    "butt",
  ],
  "_lineDashOffsetStack": Array [
    0,
    0,
  ],
  "_lineDashStack": Array [
    Array [],
    Array [],
  ],
  "_lineJoinStack": Array [
    "miter",
    "miter",
  ],
  "_lineWidthStack": Array [
    1,
    undefined,
  ],
  "_miterLimitStack": Array [
    10,
    10,
  ],
  "_path": Array [
    Object {
      "props": Object {},
      "transform": Array [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "beginPath",
    },
  ],
  "_shadowBlurStack": Array [
    0,
    0,
  ],
  "_shadowColorStack": Array [
    "rgba(0, 0, 0, 0)",
    "rgba(0, 0, 0, 0)",
  ],
  "_shadowOffsetXStack": Array [
    0,
    0,
  ],
  "_shadowOffsetYStack": Array [
    0,
    0,
  ],
  "_stackIndex": 1,
  "_strokeStyleStack": Array [
    "#000",
    "#000",
  ],
  "_textAlignStack": Array [
    "start",
    "start",
  ],
  "_textBaselineStack": Array [
    "alphabetic",
    "alphabetic",
  ],
  "_transformStack": Array [
    Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
  ],
  "addHitRegion": [MockFunction],
  "arc": [MockFunction],
  "arcTo": [MockFunction],
  "beginPath": [MockFunction],
  "bezierCurveTo": [MockFunction],
  "clearHitRegions": [MockFunction],
  "clearRect": [MockFunction],
  "clip": [MockFunction],
  "closePath": [MockFunction],
  "createImageData": [MockFunction],
  "createLinearGradient": [MockFunction],
  "createPattern": [MockFunction],
  "createRadialGradient": [MockFunction],
  "drawFocusIfNeeded": [MockFunction],
  "drawImage": [MockFunction],
  "ellipse": [MockFunction],
  "fill": [MockFunction],
  "fillRect": [MockFunction] {
    "calls": Array [
      Array [
        0,
        0,
        0,
        0,
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "fillText": [MockFunction],
  "getImageData": [MockFunction],
  "getLineDash": [MockFunction],
  "getTransform": [MockFunction],
  "isPointInPath": [MockFunction],
  "isPointInStroke": [MockFunction],
  "lineTo": [MockFunction],
  "measureText": [MockFunction],
  "moveTo": [MockFunction],
  "putImageData": [MockFunction],
  "quadraticCurveTo": [MockFunction],
  "rect": [MockFunction],
  "removeHitRegion": [MockFunction],
  "resetTransform": [MockFunction],
  "restore": [MockFunction],
  "rotate": [MockFunction],
  "save": [MockFunction] {
    "calls": Array [
      Array [],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "scale": [MockFunction],
  "scrollPathIntoView": [MockFunction],
  "setLineDash": [MockFunction],
  "setTransform": [MockFunction],
  "stroke": [MockFunction],
  "strokeRect": [MockFunction],
  "strokeText": [MockFunction],
  "transform": [MockFunction],
  "translate": [MockFunction] {
    "calls": Array [
      Array [
        0,
        0,
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
}
`;
